src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h

```diff
@@ -1708,7 +1788,176 @@ namespace UiaOperationAbstraction
         }
     }
 
-    class UiaCacheRequest: public UiaTypeBase<
+    template <class... ItemWrapperType>
+    class UiaTuple : public UiaTypeBase<
+        std::shared_ptr<std::tuple<typename ItemWrapperType::LocalType...>>,
+        winrt::Microsoft::UI::UIAutomation::AutomationRemoteArray>
+    {
+    public:
+        using TupleLocalType = std::tuple<typename ItemWrapperType::LocalType...>;
+
+        template<std::size_t I>
+        using TupleItemWrapperType = typename std::tuple_element<I, std::tuple<ItemWrapperType...>>::type;
+
+        static constexpr auto c_anyTest = &winrt::Microsoft::UI::UIAutomation::AutomationRemoteAnyObject::IsArray;
+        static constexpr auto c_anyCast = &winrt::Microsoft::UI::UIAutomation::AutomationRemoteAnyObject::AsArray;

```

Comment:
May I ask what are these (`c_anyTest`, `c_anyCast`) used for?

Reply:
Absolutely. Elements of collections (such as values of `UiaStringMap`) have requirements around type conversions from remote types to wrapper types.

For instance, `UiaStringMap` allows you to define values as wrapper types. However, the remote type of `AutomationRemoteStringMap` uses `AutomationRemoteObject` as the value type (which means it can be _almost_ any remote object).

This means that `UiaStringMap` needs to know more about the value wrapper type in order to create it out of `AutomationRemoteObject` (by casting the object to the desired remote type such as `AutomationRemoteString`, for instance). The function pointers basically help define those conversions.

`UiaStringMap` implements it in the following way:
```
        ItemWrapperType Lookup(UiaString key)
        {
            if (ShouldUseRemoteApi())
            {
                ToRemote();
                key.ToRemote();

                auto remoteAny = std::get<RemoteType>(m_member).Lookup(key);
                return std::invoke(ItemWrapperType::c_anyCast, &remoteAny);
            }

            auto localKey = key.GetLocalWstring();
            ItemLocalType result = std::get<LocalType>(m_member)->at(localKey);

            return result;
        }
```
Technically, we could probably put that on individual wrapper type constructors, but I did not want to change the current approach in any shape unless necessary to implement this change.

---
In reply to: [504924632](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/43#discussion_r504924632) [](ancestors = 504924632)

Reply:
Sure, thanks!